cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -1376,20 +1386,39 @@ private KafkaVersionChange getKafkaVersionChange(StatefulSet kafkaSts) {
          */
         Future<ReconciliationState> zkScaleUpInSmallSteps() {
             return zkSetOperations.getAsync(namespace, ZookeeperCluster.zookeeperClusterName(name))
-                    .map(sts -> sts == null ? Integer.valueOf(0) : sts.getSpec().getReplicas())
-                    .compose(currentReplicas -> {
-                        if (currentReplicas > 0 && zkCluster.getReplicas() > currentReplicas) {
-                            zkCluster.setReplicas(currentReplicas + 1);
-                            zkScalingUp = true;
-                        }
+                    .compose(sts -> {
+                        if (sts != null) {
+                            int currentReplicas = sts.getSpec().getReplicas();
+                            if (currentReplicas > 0 && zkCluster.getReplicas() > currentReplicas) {
+                                // TODO: For ZK 3.5.x we need to limit scaling to 1 at a time. Should we enforce that here?
+                                zkCluster.setReplicas(currentReplicas + 1);
+                                zkScalingUp = true;
+                            } else if (currentReplicas > 0 && zkCluster.getReplicas() < currentReplicas) {
+                                zkScalingDown = true;
+                            }
 
+                            if ((zkScalingUp || zkScalingDown) && (KafkaVersion.compareDottedVersions(zkCluster.getVersion(), "3.4.99") > 0)) {
+                                log.info(
+                                        "Scaling Zookeeper 3.5.x cluster: Rolling updates will be skipped and the manual " +
+                                                "update process defined in the documentation should be followed"
+                                );
+                                // Setting this will add the annotation to the STS in the state.zkStatefulSet() method
+                                zkManualScaling = true;
+                            } else if (Annotations.booleanAnnotation(sts, Annotations.ANNO_STRIMZI_IO_MANUAL_ZK_SCALE, false)) {
+                                // We need to check if we are still in a manual scaling operation. Only the user can set this to false
+                                // once it has been set to true.
+                                log.info("Detected manual Zookeeper scaling operation in progress");
+                                zkManualScaling = true;
+                            }
+                        }
                         return Future.succeededFuture(this);
                     });
         }
 
         Future<ReconciliationState> zkScaleUp() {
             if (zkScalingUp) {
                 return zkSetOperations.scaleUp(namespace, zkCluster.getName(), zkCluster.getReplicas())
```

Comment:
Does this or the `scaleDown` method need to be protected by the annotation as well? IIRC you said scaleup should happen node by node ... but unless this is protected, and someone changes the replicas e.g. from `3` to `7` it scale up every reocnciliation regardless whether the user finished the manual tasks, or?

Reply:
See comments above above for `zkScalingSetup` method. We would need to catch the scale down of more than one and overwrite the STS. Do we want to do that or rely on the method saying one at a time?